### ISSUE TITLE
NoDelayProvisionStrategy incorrect availability

### ISSUE THREAD

**Issue by FredrikSeidl** [jenkinsci/ec2-fleet-plugin#149]
Hi!

We have noticed that our fleets does not scale properly when using the new feature "No delay provisioning". For some reason it seems to think that we have capacity which we don't. Right now we have a Jenkins environment with one EC2-Fleet and one master. The master has 0 executors and the ec2-fleet is scaled to 2 with 2 executors per instance. Both executors are occupied and there are 2 jobs in queue and it does not add a new instance.

When looking at the logs it continues to log `Available capacity=6, currentDemand=-5` so it seems as if some state is wrong in the plugin regarding the capacity. Is it possible to reset it somehow?

**Comment by terma** [comment]
Hi,

Interesting, so log reports ```availableCapacity=6``` which means you could have some amount of free idle executors (I assume it's not the case) and some not yet connected plus some planned. From that point of view, the plugin decides to wait when new executors will appear since they will cover all demand.

Please check log if you see some exceptions or ```fleet instances:``` and add here in comments, thx

I see a couple of options:
1. Plugin decided that scale required but wasn't able to scale fleet
2. EC2 Fleet was scaled and new capacity in progress (you can check if EC2 Spot Fleet in modification state in AWS Console or Jenkins Widget shows target capacity```>2```)
  - probably need to wait a few minutes
3. EC2 Fleet was scaled and new instances were added into the fleet, you should be able to see them as disconnected in Jenkins. In that case, click on one of the node in Jenkins to see a connection log

Below code which calculate counts just in case:
```java
final int availableCapacity =
    snapshot.getAvailableExecutors()   // live executors
    + snapshot.getConnectingExecutors()  // executors present but not yet connected
    + strategyState.getPlannedCapacitySnapshot()     // capacity added by previous strategies from previous rounds
    + strategyState.getAdditionalPlannedCapacity();  // capacity added by previous strategies _this round_
```

https://github.com/jenkinsci/ec2-fleet-plugin/blob/master/src/main/java/com/amazon/jenkins/ec2fleet/NoDelayProvisionStrategy.java#L35:L41

**Comment by FredrikSeidl** [comment]
I see no exceptions in my logs and I see one instance in `fleet instances` which is the one currently connected (had to set minimum instances to 1 so our projects don't halt).
We had some issues with the scaling before because there was a launch configuration that did not work so the fleet group was set to 3 but no instances would come available. What I did then was to create a new ec2 fleet and change the configuration to that one. That adds up pretty well with the availableCapacity=6 as 3 instances * 2 executors is 6. Could it still be some reference to the old fleet that is hanging around?

**Comment by terma** [comment]
Well if you don't see any explicit exception in log, possible, I suggest to create ticket for AWS Customer Support.

**Comment by terma** [comment]
Going to close ticket, since new fleet works as expected, feel free to reopen

**Comment by FredrikSeidl** [comment]
But the EC2 fleet only has one instance and your plugin says that you have 8 executors. I don't see how this is an AWS issue? I think that it's the snapshotting logic in the plugin that remembers the 3 instances that never was available in the old fleet that we no longer have around

**Comment by terma** [comment]
my bad, I thought 

> ec2-fleet is scaled to 2 with 2 executors per instance

means two EC2 instances with two executors for each, let's summarize case:
- ```EC2 instance = 1```
- ```executors in total = 2```
- ```EC2 Spot Fleet targetCapacity = 1```
- No errors in the log
correct?

**Comment by FredrikSeidl** [comment]
Sorry for being unclear. The jenkins server is very much in use so the state differs all the time. 

`EC2 instance = 1 (master node, 0 executors)`
`EC2 Spot Fleet targetCapacity = 0`
`Actual EC2 Spot Fleet instances = 0`
`Executors per instance in Spot fleet = 2`
`Reported available capacity from plugin in logs = 8`
`Current job queue = 8`
`currentDemand = 0` 

Result of this is no scaling activity. When adding more jobs to queue it will scale but it has to go above 8 jobs in the queue in order for that to happen

**Comment by terma** [comment]
I think this one happens only when you open configuration page, plugin supports asg groups but cannot describe tgem. Not related to scale problem. Could it be xase that jobs cannot be executed on plugin nodes, labels?

**Comment by FredrikSeidl** [comment]
Ah okay.

The jobs can execute on the fleet nodes. We have no labels or limitation of that sort

**Comment by ScOut3R** [comment]
Could be related. Every now and then the build queue builds up while Jenkins refuses to spin up new workers. This has started last week when I updated the fleet plugin, I think from 2.15 to 2.16. The issue happens with the previous version as well.

I can see similar lines in the Jenkins log:
```
2019-11-05 11:16:02.983+0000 [id=36]	INFO	c.a.j.e.NoDelayProvisionStrategy#apply: Available capacity=2, currentDemand=0
2019-11-05 11:16:02.983+0000 [id=36]	INFO	c.a.j.e.NoDelayProvisionStrategy#apply: Available capacity=4, currentDemand=-2
```
Available capacity is clearly wrong, because in this scenario only one worker was provisioned with an executor cap of 1 and it was busy. There were 5 jobs waiting in line.

A Jenkins restart resolves this issue.

**Comment by terma** [comment]
Hi,

I found that no delay strategy doesn't take into account amount of executors (assumed as 1) plus added more logging for no delay provision strategy in version [1.16.1](https://github.com/jenkinsci/ec2-fleet-plugin/releases/tag/ec2-fleet-1.16.1)

Let's try this version, I hope additional log will be able to show situation. Thx.

**Comment by ScOut3R** [comment]
Thank you @terma! I've just updated the plugin. Will report back.

**Comment by FredrikSeidl** [comment]
Updating the plugin now, will report back with results 😀

**Comment by FredrikSeidl** [comment]
Initial impressions is that the scale-up works a lot better than before. However I set the minimum capacity back to 0 and noticed that my instances are still around after being idle but marked as (suspended). 

I checked the AWS console and noticed that the modify spot fleet requests look like this

`Modify request received. Requested targetCapacity: 0, excessCapacityTerminationPolicy: NoTermination`

The excessCapacityTerminationPolicy: NoTermination part seems to mean that extra servers wont be killed by AWS. The spot fleet has this value set to "default" so I'm guessing that the plugin sends this value incorrectly to AWS.

After this there wont be any scale up activity from the plugin. So the problem remains

**Comment by FredrikSeidl** [comment]
Hey @terma, have you had time to check this out? Do you need any more info from me?

**Comment by FredrikSeidl** [comment]
@terma I looked through the plugin code and found the place where the excess capacity policy is set and changed it in PR #154

**Comment by Drupal-Infrastructure** [comment]
Hi there, we're having the same /similar issues. I enabled logging for the com.amazon.jenkins.ec2fleet.NoDelayProvisionStrategy and we see the following: 

currentDemand 0 availableCapacity 8 (availableExecutors 0 connectingExecutors 0 plannedCapacitySnapshot 8 additionalPlannedCapacity 0)

In this case we had 2 running jobs, and 2 instances, and 8 jobs in the queue for a  long while.

It seems like 'plannedCapacitySnapshot' claims that its going to create new instances, but never does. Im not sure what that relates to, but its causing us to only provision if we have more jobs in the queue than we have provisioned.

**Comment by Drupal-Infrastructure** [comment]
Today we've got 15 jobs stacked up in the queue, no executors running, and the logs are saying `currentDemand 0 availableCapacity 15 (availableExecutors 0 connectingExecutors 0 plannedCapacitySnapshot 15 additionalPlannedCapacity 0)`

Whatever plannedCapacitySnapshot is, it should not be used in these calculations, or we need to figure out why it accumulates nodes without actually provisioning them.

**Comment by FredrikSeidl** [comment]
@terma is there any work going on at this issue? Right now I consider this plugin completely broken as it does not do what's intended. I've had to schedule a restart every hour to keep our builds moving forward which is not a long-term-solution

**Comment by Drupal-Infrastructure** [comment]
Howdy, getting paged on weekends and middle of the night.

9 jobs in the queue, 1 executor marked "suspended" and The logs say "currentDemand -26 availableCapacity 35 (availableExecutors 0 connectingExecutors 0 plannedCapacitySnapshot 35 additionalPlannedCapacity 0)"

**Comment by mayaaravotwm** [comment]
+1

**Comment by terma** [comment]
Fix released under version [1.16.2](https://github.com/jenkinsci/ec2-fleet-plugin/releases/tag/ec2-fleet-1.16.2)

**Comment by ryanaslett** [comment]
Fantastic!  We've installed this and nodes are now spinning up and down exactly on demand, without any issues. 

Thank you so much for solving this, it really means a lot. (I posted from the @Drupal-Infrastructure account earlier)
